### PR TITLE
Enable CA1825: Avoid zero-length array allocations

### PR DESCRIPTION
**Pull request by xtqqczze**
https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1825

Follow-up to #9042.

### CODE REVIEW

File: src/System.Management.Automation/engine/runtime/Operations/ArrayOps.cs [hunk 1]
```diff
@@ -52,7 +52,7 @@ internal static T[] Multiply<T>(T[] array, uint times)
 
             if (times == 0 || array.Length == 0)
             {
-                return new T[0]; // don't use Utils.EmptyArray, always return a new array
+                return Array.Empty<T>();
```
Comment (xtqqczze) [inline src/System.Management.Automation/engine/runtime/Operations/ArrayOps.cs:55]: @iSazonov Perhaps we should suppress `CA1825` here, if there is still a rationale for always returning a new array instance. `Utils.EmptyArray` was removed in your PR #9042.

File: docs/dev-process/coding-guidelines.md [hunk 2]
```diff
@@ -111,9 +111,6 @@ Some general guidelines:
 * Avoid using string interpolations and overloads with implicit parameters such as `Culture` and `StringComparison`.
   Instead, use overloads with more explicit parameters such as `String.Format(IFormatProvider, String, Object[])` and `Equals(String, String, StringComparison)`.
 
-* Avoid creating empty arrays.
```
Comment (xtqqczze) [inline docs/dev-process/coding-guidelines.md:114]: Unnecessary to make this guideline explicit when we can rely on the compiler to warn against this pattern.

File: src/System.Management.Automation/engine/PseudoParameters.cs [hunk 3]
```diff
@@ -236,6 +236,6 @@ public string HelpFile
         /// </summary>
         public object Data { get; set; }
 
-        internal static readonly RuntimeDefinedParameter[] EmptyParameterArray = new RuntimeDefinedParameter[0];
+        internal static readonly RuntimeDefinedParameter[] EmptyParameterArray = Array.Empty<RuntimeDefinedParameter>();
```
Comment (xtqqczze) [inline src/System.Management.Automation/engine/PseudoParameters.cs:239]: Perhaps we can inline `EmptyParameterArray `?

File: src/System.Management.Automation/engine/runtime/Operations/ArrayOps.cs (hunk 1 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/runtime/Operations/ArrayOps.cs:55]: Yes, please revert and suppress.

File: docs/dev-process/coding-guidelines.md (hunk 2 above)
Comment (iSazonov) [reply docs/dev-process/coding-guidelines.md:114]: I think we will update the guidance after we finish our current migration to Roslyn rules.

File: src/System.Management.Automation/engine/PseudoParameters.cs (hunk 3 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/PseudoParameters.cs:239]: Not in the PR.

File: src/System.Management.Automation/engine/runtime/Operations/MiscOps.cs [hunk 4]
```diff
@@ -3228,7 +3228,7 @@ internal static object Multiply(IEnumerator enumerator, uint times)
 
             if (originalList.Count == 0)
             {
-                return new object[0]; // don't use Utils.EmptyArray, always return a new array
+                return Array.Empty<object>();
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/runtime/Operations/MiscOps.cs:3231]: The same. Please revert, suppress and update comment.

File: src/System.Management.Automation/engine/runtime/Operations/ArrayOps.cs (hunk 1 above)
Comment (xtqqczze) [reply src/System.Management.Automation/engine/runtime/Operations/ArrayOps.cs:55]: I will revert and suppress, but I was wondering what reason we do not to use the static empty array?

File: docs/dev-process/coding-guidelines.md (hunk 2 above)
Comment (xtqqczze) [reply docs/dev-process/coding-guidelines.md:114]: So we'll keep this removal?

File: src/System.Management.Automation/engine/runtime/Operations/ArrayOps.cs (hunk 1 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/runtime/Operations/ArrayOps.cs]: I don't remember. Perhaps it is used in comparisons.

File: src/System.Management.Automation/engine/runtime/Operations/ArrayOps.cs [hunk 5]
```diff
@@ -52,7 +56,8 @@ internal static T[] Multiply<T>(T[] array, uint times)
 
             if (times == 0 || array.Length == 0)
             {
-                return new T[0]; // don't use Utils.EmptyArray, always return a new array
+                // Always return a new array.
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/runtime/Operations/ArrayOps.cs:59]:
```suggestion
                // Don't use Array.Empty<T>. Always return a new array.
```

File: src/System.Management.Automation/engine/runtime/Operations/MiscOps.cs [hunk 6]
```diff
@@ -3228,7 +3232,8 @@ internal static object Multiply(IEnumerator enumerator, uint times)
 
             if (originalList.Count == 0)
             {
-                return new object[0]; // don't use Utils.EmptyArray, always return a new array
+                // Always return a new array.
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/runtime/Operations/MiscOps.cs:3235]:
```suggestion
                // Don't use Array.Empty<T>. Always return a new array.

```

File: src/System.Management.Automation/engine/runtime/Operations/ArrayOps.cs [hunk 7]
```diff
@@ -41,6 +41,10 @@ internal static object[] SlicingIndex(object target, object[] indexes, Func<obje
         /// <param name="array">Collection to multiply.</param>
         /// <param name="times">Number of times the collection is to be multiplied/copied.</param>
         /// <returns>Collection multiplied by integer.</returns>
+        [System.Diagnostics.CodeAnalysis.SuppressMessage(
+            "Performance",
+            "CA1825:Avoid zero-length array allocations",
+            Justification = "Code from source-depot includes comment: 'don't use Utils.EmptyArray, always return a new array'.")]
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/runtime/Operations/ArrayOps.cs:47]: Please use local suppression. We discussed this previously.

File: src/System.Management.Automation/engine/runtime/Operations/MiscOps.cs [hunk 8]
```diff
@@ -3208,6 +3208,10 @@ internal static object PropertyGetter(PSGetMemberBinder binder, IEnumerator enum
             return result.ToArray();
         }
 
+        [System.Diagnostics.CodeAnalysis.SuppressMessage(
+            "Performance",
+            "CA1825:Avoid zero-length array allocations",
+            Justification = "Code from source-depot includes comment: 'don't use Utils.EmptyArray, always return a new array'.")]
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/runtime/Operations/MiscOps.cs:3214]: Please use local suppression. We discussed this previously.

**Review by iSazonov — APPROVED** [review]

File: src/System.Management.Automation/engine/runtime/Operations/ArrayOps.cs (hunk 1 above)
Comment (xtqqczze) [reply src/System.Management.Automation/engine/runtime/Operations/ArrayOps.cs]: @PaulHigin Must we always return a new instance, rather then using the static empty array?

**Review by PaulHigin — APPROVED** [review]

**Pull request merged**